File: CarLine.DataCleanUp/Services/Cleanup/RecordCleaner.cs
using CarLine.DataCleanUp.Models;

namespace CarLine.DataCleanUp.Services.Cleanup;

internal static class RecordCleaner
{
    public static bool CleanAndValidate(Dictionary<string, string> csvRecord, Dictionary<string, string> fullRecord)
    {
        // Validate required categorical fields - skip record if invalid
        if (!ValidateField(csvRecord, "transmission", DataCleanupConstants.ValidTransmissions))
            return false;
        if (csvRecord.TryGetValue("transmission", out var trans))
            fullRecord["transmission"] = trans;

        if (!ValidateField(csvRecord, "condition", DataCleanupConstants.ValidConditions))
            return false;
        if (csvRecord.TryGetValue("condition", out var cond))
            fullRecord["condition"] = cond;

        if (!ValidateField(csvRecord, "fuel", DataCleanupConstants.ValidFuels))
            return false;
        if (csvRecord.TryGetValue("fuel", out var fuel))
            fullRecord["fuel"] = fuel;

        if (!ValidateField(csvRecord, "type", DataCleanupConstants.ValidTypes))
            return false;
        if (csvRecord.TryGetValue("type", out var type))
            fullRecord["type"] = type;

        // Standardize manufacturer
        if (csvRecord.TryGetValue("manufacturer", out var manu) && !string.IsNullOrWhiteSpace(manu))
        {
            manu = manu.Trim().ToLowerInvariant();
            if (DataCleanupConstants.ManufacturerMap.TryGetValue(manu, out var standardized))
            {
                csvRecord["manufacturer"] = standardized;
                fullRecord["manufacturer"] = standardized;
            }
            else
            {
                csvRecord["manufacturer"] = manu;
                fullRecord["manufacturer"] = manu;
            }
        }
        else
        {
            return false;
        }

        // Model normalization
        if (csvRecord.TryGetValue("model", out var model) && !string.IsNullOrWhiteSpace(model))
        {
            var normalized = NormalizeModelName(model);
            if (string.IsNullOrWhiteSpace(normalized))
                return false;
            csvRecord["model"] = normalized;
            fullRecord["model"] = normalized;
        }
        else
        {
            return false;
        }

        // Year
        if (csvRecord.TryGetValue("year", out var yearStr) && !string.IsNullOrWhiteSpace(yearStr))
        {
            if (int.TryParse(yearStr, out var year))
            {
                if (year < 1900 || year > DateTime.UtcNow.Year + 1)
                    return false;
                csvRecord["year"] = year.ToString();
                fullRecord["year"] = year.ToString();
            }
            else
            {
                return false;
            }
        }
        else
        {
            return false;
        }

        // Price
        if (csvRecord.TryGetValue("price", out var priceStr) && !string.IsNullOrWhiteSpace(priceStr))
        {
            if (decimal.TryParse(priceStr, out var price))
            {
                if (price < 100 || price > 1000000)
                    return false;
                var normalized = price.ToString("F0");
                csvRecord["price"] = normalized;
                fullRecord["price"] = normalized;
            }
            else
            {
                return false;
            }
        }
        else
        {
            return false;
        }

        // Odometer
        if (csvRecord.TryGetValue("odometer", out var odoStr) && !string.IsNullOrWhiteSpace(odoStr))
        {
            if (int.TryParse(odoStr, out var odo))
            {
                if (odo < 0 || odo > 500000)
                    return false;

                fullRecord["odometer"] = odo.ToString();
                csvRecord["odometer"] = Math.Log10(odo + 1).ToString("F3");
            }
            else
            {
                return false;
            }
        }
        else
        {
            return false;
        }

        // Optional region
        if (csvRecord.TryGetValue("region", out var region) && !string.IsNullOrWhiteSpace(region))
        {
            var normalized = region.Trim().ToLowerInvariant();
            csvRecord["region"] = normalized;
            fullRecord["region"] = normalized;
        }

        return true;
    }

    private static bool ValidateField(Dictionary<string, string> record, string fieldName, HashSet<string> validValues)
    {
        if (!record.TryGetValue(fieldName, out var value))
            return false;

        if (string.IsNullOrWhiteSpace(value))
            return false;

        value = value.Trim().ToLowerInvariant();

        if (!validValues.Contains(value))
            return false;

        record[fieldName] = value;
        return true;
    }

    /// <summary>
    ///     Normalize model name by taking only the first word.
    /// </summary>
    private static string NormalizeModelName(string model)
    {
        if (string.IsNullOrWhiteSpace(model))
            return string.Empty;

        model = model.Trim().ToLowerInvariant();

        var firstWord = model.Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault();

        return firstWord ?? string.Empty;
    }
}

File: CarLine.DataCleanUp.FSharp/Domain/RecordCleaner.fs
module CarLine.DataCleanUp.FSharp.RecordCleaner

// Railway-Oriented Programming

open System
open System.Collections.Generic

type private Records = Dictionary<string, string> * Dictionary<string, string>

let private validateField fieldName (validValues: HashSet<string>) ((csv, full): Records) =
    match csv.TryGetValue(fieldName) with
    | true, v when not (String.IsNullOrWhiteSpace(v)) ->
        let norm = v.Trim().ToLowerInvariant()
        if validValues.Contains(norm) then
            csv[fieldName]  <- norm
            full[fieldName] <- norm
            Ok (csv, full)
        else Error $"Invalid {fieldName}: '{norm}'"
    | _ -> Error $"Missing {fieldName}"

let private standardizeManufacturer ((csv, full): Records) =
    match csv.TryGetValue("manufacturer") with
    | true, manu when not (String.IsNullOrWhiteSpace(manu)) ->
        let norm = manu.Trim().ToLowerInvariant()
        let standardized =
            match DataCleanupConstants.manufacturerMap.TryGetValue(norm) with
            | true, mapped -> mapped
            | _            -> norm
        csv["manufacturer"]  <- standardized
        full["manufacturer"] <- standardized
        Ok (csv, full)
    | _ -> Error "Missing manufacturer"

let private normalizeModel ((csv, full): Records) =
    match csv.TryGetValue("model") with
    | true, model when not (String.IsNullOrWhiteSpace(model)) ->
        let firstWord =
            model.Trim().ToLowerInvariant()
                 .Split([| ' '; '\t' |], StringSplitOptions.RemoveEmptyEntries)
            |> Array.tryHead
        match firstWord with
        | Some word ->
            csv["model"]  <- word
            full["model"] <- word
            Ok (csv, full)
        | None -> Error "Empty model after normalization"
    | _ -> Error "Missing model"

let private validateYear ((csv, full): Records) =
    match csv.TryGetValue("year") with
    | true, s when not (String.IsNullOrWhiteSpace(s)) ->
        match Int32.TryParse(s) with
        | true, y when y >= 1900 && y <= DateTime.UtcNow.Year + 1 ->
            let str = string y
            csv["year"]  <- str
            full["year"] <- str
            Ok (csv, full)
        | true, _ -> Error "Year out of range"
        | _       -> Error "Invalid year"
    | _ -> Error "Missing year"

let private validatePrice ((csv, full): Records) =
    match csv.TryGetValue("price") with
    | true, s when not (String.IsNullOrWhiteSpace(s)) ->
        match Decimal.TryParse(s) with
        | true, p when p >= 100m && p <= 1_000_000m ->
            let norm = p.ToString("F0")
            csv["price"]  <- norm
            full["price"] <- norm
            Ok (csv, full)
        | true, _ -> Error "Price out of range"
        | _       -> Error "Invalid price"
    | _ -> Error "Missing price"

let private validateOdometer ((csv, full): Records) =
    match csv.TryGetValue("odometer") with
    | true, s when not (String.IsNullOrWhiteSpace(s)) ->
        match Int32.TryParse(s) with
        | true, odo when odo >= 0 && odo <= 500_000 ->
            full["odometer"] <- string odo
            csv["odometer"]  <- Math.Log10(float odo + 1.0).ToString("F3")
            Ok (csv, full)
        | true, _ -> Error "Odometer out of range"
        | _       -> Error "Invalid odometer"
    | _ -> Error "Missing odometer"

let private normalizeRegion ((csv, full): Records) =
    match csv.TryGetValue("region") with
    | true, region when not (String.IsNullOrWhiteSpace(region)) ->
        let norm = region.Trim().ToLowerInvariant()
        csv["region"]  <- norm
        full["region"] <- norm
    | _ -> ()
    Ok (csv, full)

let cleanAndValidate (csvRecord: Dictionary<string, string>) (fullRecord: Dictionary<string, string>) : bool =
    Ok (csvRecord, fullRecord)
    |> Result.bind (validateField "transmission" DataCleanupConstants.validTransmissions)
    |> Result.bind (validateField "condition"    DataCleanupConstants.validConditions)
    |> Result.bind (validateField "fuel"         DataCleanupConstants.validFuels)
    |> Result.bind (validateField "type"         DataCleanupConstants.validTypes)
    |> Result.bind standardizeManufacturer
    |> Result.bind normalizeModel
    |> Result.bind validateYear
    |> Result.bind validatePrice
    |> Result.bind validateOdometer
    |> Result.bind normalizeRegion
    |> Result.isOk
